src/com/facebook/buck/jvm/java/JarShape.java

```diff
@@ -0,0 +1,133 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.jvm.java;
+
+import com.facebook.buck.graph.AbstractBreadthFirstTraversal;
+import com.facebook.buck.rules.BuildRule;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public enum JarShape {
+  MAVEN {
+    @Override
+    public Summary gatherDeps(BuildRule root) {
+      // Our life is a lot easier if the root is a JavaLibrary :)
+      if (!(root instanceof JavaLibrary)) {
+        throw new RuntimeException("Boom");
+      }
+
+      // We only need the first order maven deps, since maven's depenedency resolution process will
+      // pull in any transitive deps. Gather candidate maven deps using a BFS
+      Set<HasMavenCoordinates> mavenDeps = new HashSet<>();
+      Set<JavaLibrary> combinedTransitiveDeps = new HashSet<>();
+      new AbstractBreadthFirstTraversal<JavaLibrary>((JavaLibrary) root) {
+        @Override
+        public Iterable<JavaLibrary> visit(JavaLibrary javaLibrary) throws RuntimeException {
+          if (!javaLibrary.equals(root) && javaLibrary.getMavenCoords().isPresent()) {
+            mavenDeps.add(javaLibrary);
+            // We may have a transitive dep on another maven dep.
+            combinedTransitiveDeps.addAll(
+                javaLibrary.getTransitiveClasspathDeps().stream()
+                    .filter(input -> !(input.equals(javaLibrary)))
+                    .filter(input -> input.getMavenCoords().isPresent())
+                    .collect(Collectors.toSet()));
+            return ImmutableSet.of();
+          }
+          return javaLibrary.getTransitiveClasspathDeps();
+        }
+      }.start();
+      mavenDeps.removeAll(combinedTransitiveDeps);
+
+      ImmutableSet<JavaLibrary> classpath = ((HasClasspathEntries) root).getTransitiveClasspathDeps();
+      Set<JavaLibrary> toPackage = classpath.stream()
+          .filter(input -> !combinedTransitiveDeps.contains(input))
+          .filter(input -> !mavenDeps.contains(input))
+          .collect(Collectors.toSet());
+
+      return new Summary(
+          toPackage,
+          classpath,
+          mavenDeps);
+    }
+  },
+  SINGLE {
+    @Override
+    public Summary gatherDeps(BuildRule root) {
+      ImmutableSet<JavaLibrary> classpath;
+      if (root instanceof HasClasspathEntries) {
+        classpath = ((HasClasspathEntries) root).getTransitiveClasspathDeps();
+      } else {
+        classpath = ImmutableSet.of();
+      }
+
+      return new Summary(
+          ImmutableSortedSet.of(root),
+          classpath,
+          ImmutableSortedSet.of());
+    }
+  },
+  UBER {
```

Comment:
This is not even used yet, right? Could we add it in the commit that actually needs it?